Userland/Utilities/file.cpp

```diff
@@ -51,6 +52,39 @@ static Optional<DeprecatedString> image_details(StringView description, StringVi
     return builder.to_deprecated_string();
 }
 
+static Optional<DeprecatedString> audio_details(StringView description, StringView path)
+{
+// If serenity_main is defined as a macro, we're compiling for BuggieBox.
+#if defined(serenity_main)
+    return description_only(description, path);
+#else
+    auto loader_or_error = Audio::Loader::create(path);
+    if (loader_or_error.is_error())
+        return {};
+
+    auto loader = loader_or_error.release_value();
+    StringBuilder builder;
+    builder.appendff("{}, {} Hz, {}-bit {}, {} samples ({} s)\n",
+        description,
+        loader->sample_rate(),
+        loader->bits_per_sample(),
+        loader->num_channels() == 1 ? "Mono" : "Stereo",
+        loader->total_samples(),
+        loader->total_samples() / loader->sample_rate());
+    auto metadata = loader->metadata();
+    if (metadata.title.has_value()) {
+        // FIXME: Use “pretty quotation” once our terminal fonts support these characters.
+        builder.appendff("\"{}\"", metadata.title.value_or({}));
+    }
+    if (metadata.album.has_value())
+        builder.appendff(" (Album: {})", metadata.album.value_or({}));
+    if (auto all_artists = metadata.all_artists(); !all_artists.is_error() && metadata.first_artist().has_value())
+        builder.appendff(" by {}", all_artists.release_value());
```

Comment:
This is still an awkward API. We pull all artists, then make a separate API check for the first artist to see if the all-artists result has something we want? Either:

A. Make all_artists tell you there aren't any artist, e.g. with `Optional` like I said before. I disagree with your point about `all_artists` needing to handle default fallback text. LibAudio is too low level to be concerned with what text a GUI program wants to display when there are no artists - it shouldn't have a fallback param at all.

B. Make `all_artists` return a `Vector` and let the caller decide how/if to join them. I think I prefer this one actually. `all_artists` absolutely does not need to be so generic, when every caller is already giving it different fallback options / concatenation separators.

Reply:
> B. Make `all_artists` return a `Vector` and let the caller decide how/if to join them. I think I prefer this one actually. `all_artists` absolutely does not need to be so generic, when every caller is already giving it different fallback options / concatenation separators.

I disagree; at that point I would be reimplementing a very similar function for three programs. For example, there's the bracketing special case that this function handles. The vector API you suggest is as simple as copying `metadata.artists` and calling `remove_all_matching(person => !person.is_artist())` on the new vector.

I think the Optional solution is more appropriate, since you're right that a lot of users will special-case the "no artists" situation.

Reply:
> The vector API you suggest is as simple as copying metadata.artists and calling remove_all_matching(person => !person.is_artist()) on the new vector.

And? Simple > awkward.

```c++
ErrorOr<Vector<String>> Metadata::all_artists() const
{
    Vector<String> artists;

    for (auto const& person : people) {
        if (!person.is_artist())
            continue;

        if (auto role_name = person.name_for_role(); role_name.has_value())
            TRY(artist_texts.try_append(TRY(String::formatted("{} ({})", person.name, role_name.release_value()))));
        else
            TRY(artist_texts.try_append(person.name));
    }

    return artists;
}

static Optional<DeprecatedString> audio_details(StringView description, StringView path)
{
    builder.append(" by "sv);
    if (auto all_artists = metadata.all_artists().release_value_but_fixme_should_propagate_errors(); all_artists.is_empty()) 
        builder.join(", "sv, all_artists);
     else
        builder.append("Unknown"sv);
}

void SoundPlayerWidgetAdvancedView::file_name_changed(StringView name)
{
    if (auto all_artists = metadata.all_artists().release_value_but_fixme_should_propagate_errors(); all_artists.is_empty()) {
        builder.appendff("{} - ", metadata.title.value());
        builder.join(" / "sv, all_artists);
        title = builder.to_deprecated_string();
    }
}
```

Reply:
I just see the problem that I'll duplicate this code (or part of it) again for the playlist view as well as aplay. And, after all, file already needs the same logic, just with slightly different separators. I don't see a benefit of not writing this in a DRY way.

Reply:
What, exactly, do you think you'll be repeating? Specifying the same separator / a similar looking call to `StringBuilder::join`? Creating strange APIs to avoid a sliver of repetition is being *too* DRY.

Reply:
I will be repeating all of the bracketing special casing, the filtering, and the join call. We have plenty of functions that factor out this kind of code just for one or two users, I really don't see your point here.

Reply:
> all of the bracketing special casing
What bracketing?

> the filtering
No? The filtering for artists still happens inside all_artists.

> the join call
The one thing being repeated, and it's not any more than you'd be repeating otherwise. Either you repeat an `append` call or a `join` call.

My point is that this API has been awkward on every iteration so far, from first needing some "\e\e\e" string to detect no artists. But the current iteration isn't as bad IMO, so sure, let's go with it.